guides/md/writing_quantization_compatible_layers.md

```diff
@@ -0,0 +1,612 @@
+        input_dim = input_shape[-1]
+        self._kernel = self.add_weight(
+            name="kernel",
+            shape=(input_dim,),
+            initializer="random_uniform",
+        )
+
+    def call(self, inputs, training=None):
+        return ops.multiply(inputs, self._kernel)
+
+```
+
+### The `quantize()` method
+
+PTQ is a one-time rewrite. After you train or load your FP32 layer, you call
+`layer.quantize("int8")`. The layer should:
+
+1. Read its existing full-precision variables (e.g., `self._kernel`).
+2. Quantize them to INT8 values plus a quantization scale.
+3. Replace full-precision variables with INT8 storage and assign the quantized
+  data.
+4. Switch the `dtype_policy` to a quantized variant (e.g., `int8_from_float32`).
+
+
+```python
+
+def quantize(self, mode, **kwargs):
+    if mode != "int8":
+        raise NotImplementedError(f"Unsupported quantization mode: {mode}")
+
+    quantized_kernel, scale = quantizers.abs_max_quantize(
+        self._kernel, axis=0, to_numpy=True
+    )
+    scale = ops.squeeze(scale, axis=0)
+
+    kernel_shape = self._kernel.shape
+
+    del self._kernel
+
+    # Allocate INT8 variables. Discussed in the next section.
+    self._int8_build(kernel_shape)
+
+    self._kernel.assign(quantized_kernel)
+    self.scale.assign(scale)
+
+    # `_is_quantized` should be set before changing dtype policy to inform
+    # the setter that quantized variables are initialized.
+    self._is_quantized = True
+
+    if self.dtype_policy.quantization_mode is None:
+        policy = dtype_policies.get(f"{mode}_from_{self.dtype_policy.name}")
+        self.dtype_policy = policy
+
+```
+
+#### Note
+
+1. The `quantize(...)` method should validate `mode` and raise a
+  `NotImplementedError` if the mode is not supported.
+2. Ensure your `quantize(...)` sets a quantized dtype policy based on the
+  prior policy, e.g., `int8_from_float32` or `int8_from_bfloat16`. This ensures
+  that the layer's `quantization_mode` is correctly set.
+
+3. The `_is_quantized` flag should be set before changing the dtype policy to
+  inform the setter that quantized variables are initialized.
+
+### The `_int8_build(...)` method
+
+This `int8_build(...)` method is called from `quantize(...)` to initialize the
+INT8 variables. It should allocate:
+
+- `self._kernel` as an INT8 vector of shape `(input_dim,)` (the same shape as
+  the original full-precision kernel).
+- `self.scale` as the scalar quantization scale in FP32.
+
+
+```python
+
+def _int8_build(self, kernel_shape):
+    (input_dim,) = kernel_shape
+    self._kernel = self.add_weight(
+        name="kernel",
+        shape=(input_dim,),
+        initializer="zeros",
+        dtype="int8",
+        trainable=False,
+    )
+    self.scale = self.add_weight(
+        name="scale",
+        initializer="ones",
+        trainable=False,
+    )
+
+```
+
+#### Note
+
+1. The INT8 variables should be `trainable=False` since PTQ does not involve
+  further training.
```

Comment:
What happens if you make them trainable and do `model.fit()` do you get an error?

Doing `trainable=False` is a bit of a double edge sword. If you do train, what will happen is that it will only train non-quantized layers, which could be what you want, but maybe not.

Reply:
Returning an error might not be the right thing to do, since that might break the fine-tuning (for non-quantized layers) and QLoRA (for quantized layers) flow. I've modified the comment to be clearer that it's the quantized variables that should no longer be modified, and not the rest of the model.

Thanks for catching this.